Extending Isofit to handle multisurface classifications

evan-greenbrg opened the pull request:
This is includes changes to the code base to handle an external classification file that is mapped to specific configurations of surface models and executions varying per-pixel or per-superpixel. 

**The goals of the changes:**
1. Utilize external or internal classifications of surface configurations via apply_oe, directly in a manually configured config, and able to run both in per-pixel mode and in superpixel modes.

2. Remain efficient both in terms of computation time and memory requirements.

**Things to highlight:**
1. New backwards compatible "Surface" section of the isofit config. Based on the input or generated classification file, the surface config is populated.

```        
"surface": {
    "Surfaces": {
        "glint_model_surface": {
            "surface_category": "glint_model_surface",
            "surface_file": "/home/bgreenbe/store/Projects/Biosscape/prism/study_area/prm20231110t071521/Multisurface/027/data/surface_glint_model_surface.mat",
            "surface_int": 0
         },
         "multicomponent_surface": {
            "surface_category": "multicomponent_surface",
            "surface_file": "/home/bgreenbe/store/Projects/Biosscape/prism/study_area/prm20231110t071521/Multisurface/027/data/surface_multicomponent_surface.mat",
                    "surface_int": 1
                }
            },
            "base_surface_class_file": "/home/bgreenbe/store/Projects/Biosscape/prism/study_area/prm20231110t071521/Multisurface/027/input/prm20231110t071521_surface_class",
            "multi_surface_flag": true,
            "surface_class_file": "/home/bgreenbe/store/Projects/Biosscape/prism/study_area/prm20231110t071521/Multisurface/027/input/prm20231110t071521_subs_surface_class"
        }
}
```

The "Surfaces" key is only used when running multiple surfaces.

Now surface specific surface.mat files are possible, and encouraged. When generating from .json, the surface model will use a "surface type" key to generate shared surface groupings into respective .mat files. 

The cleanest implementation was to first make this a **required key for multi-surface runs from .json**.  This could be made optional. Passed .mat files will use the same surface prior file across types.

```
      {
        "input_spectrum_files":
          [
            "/home/bgreenbe/store/Data/SurfaceLibraries/filtered_other"
          ],
        "n_components": 1,
        "windows": [
          {"interval":[300,740], "regularizer":10, "correlation":"decorrelated"},
          {"interval":[740,1250], "regularizer":1e-6, "correlation":"EM", "name":"shallow-water"},
          {"interval":[1250,1325], "regularizer":1e-8, "correlation":"EM", "name": "osf"},
          {"interval":[1325,1960], "regularizer":10, "correlation": "decorrelated" },
          {"interval":[1960,2070], "regularizer":1e-6, "correlation":"EM","name": "co2" },
          {"interval":[2070,2300], "regularizer":10, "correlation":"decorrelated" },
          {"interval":[2300,2500], "regularizer":1e-3, "correlation":"EM",  "isolated": 1,"name": "noise" }
        ],
        "surface_type": "multicomponent_surface"
      },
```

2. Isofit is now run sequentially across surface types. 

```
surface_index = index_spectra_by_surface(input_config, index_pairs)
for surface_class_str, class_idx_pairs in surface_index.items():
    logging.info(f"Running surfaces: {surface_class_str}")
    ...
    # If multisurface, update config to reflect surface.
    # Otherwise, returns itself
    config = update_config_for_surface(
        deepcopy(input_config), surface_class_str
    )
    self.fm = fm = ForwardModel(config)
    ...
```

3. On the spectrum write, the statevector is matched against a full statevector by name (not index). A fill value populates elements that are present in the full statevector, but not a pixel statevector.

4. Analytical line processing is handled similarly in sequence across present surface types

A couple of other changes to note:

- Refactored the analytical line. I simplified the file initialization, which reduced clutter updating output metadata. Would like a second pair of eyes on this to make sure the read/write is similarly memory intensive as dev branch. 
- Tests for multisurface runs (Would need to hook up to github actions)
- Built in classifier based on the surface.json file. So I didn't have to build an external classification pipeline I incorporated a classifier based on the surface prior selection.
- Generalized extractions.py to use "reducers." Extractions can now take any generic reducer to move between full pixel and super pixel resolution. Necessary for aggregating classified pixels.
- Misc cleaning throughout.

There are a couple glint-related updates that I'm going to include in a separate PR.

unbohn commented on isofit/radiative_transfer/radiative_transfer.py:
```diff
@@ -81,6 +82,18 @@ def __init__(self, full_config: Config):
 
+            """Handle glint model flag from the surface model and
+            the topography model for a multipart-transmittance lut
+            Should be depreciated with #524?
```
Yes, correct! With #524, ISOFIT will be able to obtain any surface specific information from the surface model itself without the need to set a flag in the config file. In addition, the topography model will be applied automatically as soon as the user provides a LUT with separated transmittance quantities, so no flag required here either.

unbohn commented on isofit/utils/template_construction.py:
```diff
@@ -1686,3 +1719,115 @@ def reassemble_cube(matching_indices: np.array, paths: Pathnames):
+                "nonwater": "multicomponent_surface",
+                "cloud": "multicomponent_surface",
+                "uniform_surface": "multicomponent_surface",
+            }
```
@evan-greenbrg What are `nonwater` and `uniform_surface`? Aren't they covered by the `land `key?

evan-greenbrg replied:
If not provided, the surface_mapping is hard-coded so that the program can read the classification name from the file and match it with the Isofit surface name. This means that the user doesn't have to use the Isofit surface names when constructing the classification image. 

Because there is some less than ideal specificity that is required, I've used keys here that include multiple potential class names for a given surface model (e.g. water and glint as well as land and nonwater). All keys in the surface_mapping object aren't used in the processing, just the ones present in the image.

I have an associated Jupyter Notebook and data cube that helps explain the set up for the classification file. I need to work on merge in into Isotuts and get the data cube into a place where it can be downloaded.

unbohn replied:
That's a great idea! Having a tutorial notebook for this topic would be really helpful.

unbohn commented:
@evan-greenbrg Thanks for getting this fresh version of the multisurface classification up! This is really awesome and pushes ISOFIT to the next level. However, I'm currently testing the compatibility of your updates with existing frameworks and encounter one issue/question. For inspecting single pixels/inversions, I used to run the following code block:

```
from isofit.configs import configs
from isofit.core import isofit
from isofit.core.forward import ForwardModel
from isofit.inversion.inverse import Inversion

config = configs.create_new_config(
    config_file="path_to_ISOFIT_main_config")
fm = ForwardModel(config)
inv = Inversion(config, fm)
workers = isofit.Worker(config, fm, loglevel="INFO", logfile=None)
input_data = workers.io.get_components_at_index(row, col)
res = workers.iv.invert(input_data.meas, input_data.geom)
```

With your updates, the Worker class now requires two more positional arguments: `full_statevector` and `total_samples`. Is there an easy way to provide these? Or does it basically mean that the above code snippet is not executable anymore?

evan-greenbrg commented:
Thanks so much @unbohn for taking a look at this so quickly. Per your above comment, as currently written you are correct that the snippet is not currently executable. The Worker class requires the two missing arguments, which doesn't raise an issue when running the Worker via Isofit.run. 

I don't think it would be hard at all to make the missing positional arguments optional keyed arguments. `total_samples` is just used for logging and can be default to 1. `full_statevector` we can pull from the fm class if it's not provided. Would that be a useful edit?

unbohn commented:
Yes, thanks @evan-greenbrg. That sounds like a reasonable edit. I'll test and report back if I run into other issues.

evan-greenbrg committed:
Added Multicomponent classifier and fixed Analytical Line

Initializing output matrices in analytical line with numpy not envi

Fixed classification cleaning function

Fixed index shift in multicomponent classifier

Analytical Line bug fixes

Adding coupled total transm to hueristic atmosphere

Removing duplciate analytical_model function from rebase

Fixed typo in template_constructions

Changing simple to algebraic

Fixed logging

Changed state order to alphabetical

Bad object reference

Rolling back transm change

Fixing bugs in examples test

Fixing rebase issues

Fixed examples

Fixed bad invert_algebraic return

Typo in template

Fixed bad classification filtering

evan-greenbrg committed:
Edited sky glint prior, added indexing in glint model, edited bands for fit_params

Updates to AL and sky glint prior

Edits to sky glint prior

Editing glint fit parameters and adding indexing

Simplifying fit_params wl choice

Fixing typoes in surface_glint_model

Rebase with dev

evan-greenbrg committed:
Can now set surface.mat to a specific surface model. Added shared output init func.

Missed incorrect function argument

evan-greenbrg committed:
Added unit tests for multisurface runs

Loose bugs to get running through AOE

Removing multi-scattering from AOE. Edits to AOE init to remove discontinuities

Glint updates to heuristic atm

Fix surface path in config and correct surface creation on non multisurface runs

evan-greenbrg committed:
Fixing incorrect background states passing into calc_rfl (for glint)

Fixing classification cli. Tweaking glint est on init

Fixed staging and config for single surface json

Moving multistate to core

Moved multistate.py to core. Fixed surface config for input .mat files.

Fixing direct .mat input for surface

Bad indent

Missed surface_mapping var

pgbrodrick commented on isofit/utils/template_construction.py:
```diff
@@ -476,17 +522,45 @@ def check_surface_model(surface_path: str, wl: np.array, paths: Pathnames) -> st
                 "No surface model provided. Build new one using given config file."
             )
             surface_model(
-                config_path=surface_path, wavelength_path=paths.wavelength_path
+                config_path=surface_path,
+                wavelength_path=paths.wavelength_path,
+                multisurface=multisurface,
             )
             configdir, _ = os.path.split(os.path.abspath(surface_path))
```
Move down below, and write to the specified output path.

evan-greenbrg commented on isofit/utils/template_construction.py:
```diff
@@ -476,17 +522,45 @@ def check_surface_model(surface_path: str, wl: np.array, paths: Pathnames) -> st
             logging.info(
                 "No surface model provided. Build new one using given config file."
             )
             surface_model(
```
Move the  create surface model down, use the output_file flag to make sure that if I'm generating the surface model within apply oe, I'm always outputing to data directory.

pgbrodrick commented on isofit/utils/apply_oe.py:
```diff
@@ -20,7 +20,13 @@
+    extractions,
+    interpolate_spectra,
+    multicomponent_classification,
+    reducers,
```
Seems like it might be overloaded.

pgbrodrick commented on isofit/utils/template_construction.py:
```diff
@@ -1821,3 +1897,105 @@ def write_wavelength_file(filename, wl, fwhm):
+
+        # Get the class mapping. Tried to build in some insensitivity here.
+        # Will use first option it hits
+        options = [
+            "mapping",
+            "Mapping",
+            "class names",
+            "Class names",
+            "class",
+            "Class",
+            "surfaces",
+            "Surfaces",
+            "surface names",
+            "Surface names",
+        ]
+        class_mapping = None
+        for option in options:
+            if class_mapping:
+                continue
```
Per discussion, the suggestion is to enforce a standardized dictionary mapping between given index values and particular surface state types.

evan-greenbrg commented on isofit/core/isofit.py:
```diff
@@ -129,64 +143,117 @@ def run(self, row_column=None):
+            )
+
+            # Set forward model
+            self.fm = fm = ForwardModel(config)
```
Get rid of self.fm here. Could also work some cacheing into with the RT portion to try not to have to reload the LUT each iteration.

evan-greenbrg committed:
Revisions based on group discussion

Fixing bugs after rebase

isofit run gets n_rows from SpectrumFile

Moving tests into test_examples

editing working directory in apply oe test

Edits to test_cli

SpectrumFile errors out if trying to build output that already exists

Bug fix for tests

Added initialization method to IO class

Removing multisurface test label

updaing flush dirs to reflect correct working dir

Update Glint model. Fix surface copying

Surface matrices size checks for 1-component surfaces

Fixed typos

Fixed logging type

evan-greenbrg commented:
@pgbrodrick / @unbohn 

I've updated the PR with the following revisions based on our conversation:

Major changes:
- [ ] Hard-code mapping between surface class file int and surface_category. Remove all fuzzy logic around these mappings.
- [ ] Revise copy logic for surface model. If passing .json, make sure we're not generating additional .mat file. Always use data folder when running via Apply OE. Any surface copying file should _only_ happen if passing `.mat` file.
- [ ] Get run times to back up indexing strategy (see comment above).
- [ ] Work caching into the RT load. Pass in pre-init RT object into the forward model as an optional argument.
- [ ] Clean up the shared memory space in the workers. Only pass in large objects.

Minor changes:

- [ ] Call surface types key something else. 
- [ ] Add logical check in check_surface to make sure dict keys are present in .mat file if multisurface is run.
- [ ] Change name of reducers, may be built-in name. Reducers itself is not a standard library Python variable name. reduce(), however is. I've decided to keep reducers here because it's the most accurate term for what the functions do.
- [ ] Force any mapping that does exist to be alphebetical
- [ ] Rip out cloud portion of surface config
- [ ] Return all indexes from construct_full_state()
- [ ] Move match_statevector into multistate
- [ ] Pull out delete keys from initialize function
- [ ] Check all file closures.
- [ ] Make sure I'm not calling extractions twice in Apply OE
- [ ] Don't change the initialization to IO, just initialize later

evan-greenbrg commented on isofit/utils/analytical_line.py:
```diff
@@ -295,33 +394,65 @@ def __init__(
+            )
+        else:
+            self.svf = []
+
```
@brentwilder I ended making the skyview file consistent with the rest of the input files in that they are held as memory objects at the Worker level. 

I also edited the case of no sky-view file to get around holding the ones array in memory. Let me know if you have any reservations at all.

It'd be super useful for you to run through  a sky-view case on this PR whenever you have the bandwidth to do so!

unbohn commented on isofit/utils/apply_oe.py:
```diff
@@ -426,24 +426,25 @@ def apply_oe(
+            surface_working_path = f"{name}{ext}"
+
+        if value != surface_path and surface_path.endswith(".mat"):
+            copyfile(value, surface_working_path)
```
@evan-greenbrg I'm currently testing the multisurface branch (including backwards compatibility) and running into a `FileNotFoundError: [Errno 2] No such file or directory: '/work/09147/bohn/ls6/EMIT_priors/ISOFIT_dev_processing/emit20250327T212148_2508614_006_L2A_multisurface/data/surface.mat'` likely introduced by this change. It seems like ISOFIT is still expecting the surface.mat file to be located within the working directory, but with the above statement it did not copy over.

unbohn replied:
Would `==` instead of `!=` do it in line 446? `surface_path` is the argument provided to apply_oe here and `value` comes out of `tmpl.check_surface_model()`, and should be equal to `surface_path` in case we're providing a `.mat` file. This would fulfill the list item "any surface copying file should only happen if passing .mat file."

evan-greenbrg replied:
It depends on what we want to be the default behavior to be. The line in question was pointing to the wrong `surface_path` variable .

It should have been:
```python
        if value != surface_working_path and surface_path.endswith(".mat"):
            copyfile(value, surface_working_path)
```

With that said, this is under the assumption that the default behavior is to copy all external paths into the data folder. If we want the default behavior to be to not copy the file over we would do something like:

```python
        paths.surface_working_paths[key] = value
```

Do you have a preference?

unbohn replied:
My preference would be to copy all external files into the data folder under the working directory that gets passed to apply_oe. That way, it'd be easier to keep overview about the used files for a specific run.

unbohn commented on isofit/utils/multicomponent_classification.py:
```diff
@@ -0,0 +1,455 @@
+            model_dict = surface_model_dict
+        else:
+            for key in keys_to_combine:
+                model_dict[key] = np.concatenate(
```
@evan-greenbrg When running apply_oe with the below `.json` file as `--surface_path`, I'm getting a `ValueError: all the input arrays must have same number of dimensions, but the array at index 0 has 2 dimension(s) and the array at index 1 has 3 dimension(s)` here. Apparently, `model_dict[key]` has shape (285,285) and `surface_model_dict[key]` has shape (6,285,285) for the key `covs`.

```
{
  "output_model_file": "surface_20240103_multisurface.mat",
  "wavelength_file":   "emit_wavelength_fit.txt",
  "normalize":"Euclidean",
  "reference_windows":[[400,1250],[1500,1700],[2100,2450]],
  "sources":
    [
      {
        "input_spectrum_files":
          [
            "filtered_other"
          ],
        "n_components": 1,
        "windows": [
          {"interval":[300,740], "regularizer":10, "correlation":"decorrelated"},
          {"interval":[740,1250], "regularizer":1e-6, "correlation":"EM", "name":"shallow-water"},
          {"interval":[1250,1325], "regularizer":1e-8, "correlation":"EM", "name": "osf"},
          {"interval":[1325,1960], "regularizer":10, "correlation": "decorrelated" },
          {"interval":[1960,2070], "regularizer":1e-6, "correlation":"EM","name": "co2" },
          {"interval":[2070,2300], "regularizer":10, "correlation":"decorrelated" },
          {"interval":[2300,2500], "regularizer":1e-3, "correlation":"EM",  "isolated": 1,"name": "noise" }
        ],
        "surface_category": "multicomponent_surface"
      },
      {
        "input_spectrum_files":
          [
            "surface_mixture_veg_soil"
          ],
        "n_components": 2,
        "windows": [
          {"interval":[300,740], "regularizer":1e-4, "correlation":"decorrelated"},
          {"interval":[740,1250], "regularizer":1e-6, "correlation":"EM", "name": "shallow-water"},
          {"interval":[1250,1325], "regularizer":1e-8, "correlation":"EM", "name": "osf"},
          {"interval":[1325,1960], "regularizer":10, "correlation": "decorrelated"},
          {"interval":[1960,2070], "regularizer":1e-4, "correlation":"EM","name": "co2"},
          {"interval":[2070,2470], "regularizer":1e-3, "correlation":"decorrelated"},
          {"interval":[2470,2500], "regularizer":1e-6, "correlation":"EM","isolated": 1, "name": "noise" }
        ],
        "surface_category": "multicomponent_surface"
      },
      {
        "input_spectrum_files":
          [
            "filtered_veg"
          ],
        "n_components": 1,
        "windows": [
          {"interval":[300,350], "regularizer":10, "correlation":"decorrelated"},
          {"interval":[350,500], "regularizer":1e-5, "correlation":"EM","name":"aerosol","feather_forward":1e-3},
          {"interval":[500,740], "regularizer":1e-4, "correlation":"decorrelated"},
          {"interval":[740,1250], "regularizer":1e-6, "correlation":"EM", "name": "shallow-water"},
          {"interval":[1250,1325], "regularizer":1e-8, "correlation":"EM", "name": "osf"},
          {"interval":[1325,1960], "regularizer":10, "correlation": "decorrelated"},
          {"interval":[1960,2070], "regularizer":1e-4, "correlation":"EM","name": "co2","feather_forward":2e-3},
          {"interval":[2070,2470], "regularizer":1e-3, "correlation":"decorrelated"},
          {"interval":[2470,2500], "regularizer":1e-6, "correlation":"EM","isolated": 1, "name": "noise" }
        ],
        "surface_category": "multicomponent_surface"
      },
      {
        "input_spectrum_files":
          [
            "filtered_ocean"
          ],
        "n_components": 1,
        "windows": [
          {"interval":[300,880], "regularizer":10, "correlation":"decorrelated"},
          {"interval":[880,1250], "regularizer":1e-6, "correlation":"EM", "name": "shallow-water"},
          {"interval":[1250,1325], "regularizer":1e-6, "correlation":"EM", "name": "osf"},
          {"interval":[1325,2500], "regularizer":10, "correlation": "EM" }
        ],
        "surface_category": "glint_model_surface"
      },
      {
        "input_spectrum_files":
          [
            "surface_Liquids"
          ],
        "n_components": 2,
        "windows": [
          {"interval":[300,880], "regularizer":10, "correlation":"decorrelated"},
          {"interval":[880,1250], "regularizer":1e-6, "correlation":"EM", "name": "shallow-water"},
          {"interval":[1250,1325], "regularizer":1e-6, "correlation":"EM", "name": "osf"},
          {"interval":[1325,1960], "regularizer":10, "correlation": "decorrelated" },
          {"interval":[1960,2070], "regularizer":1e-4, "correlation":"EM","name": "co2","feather_forward":1e-3},
          {"interval":[2070,2470], "regularizer":1e-4, "correlation":"decorrelated"},
          {"interval":[2470,2500], "regularizer":1e-6, "correlation":"EM","isolated": 1, "name": "noise" }
        ],
        "surface_category": "multicomponent_surface"
      }
  ]
}

```

evan-greenbrg replied:
Good catch. Looks like a bug from using a single component for the filtered_ocean surface.

unbohn commented on isofit/utils/multicomponent_classification.py:
```diff
@@ -185,6 +185,9 @@ def load_surface_mat(
         else:
             for key in keys_to_combine:
+                assert (
+                    model_dict[key].ndims == surface_model_dict[key].ndims
```
@evan-greenbrg I guess this needs to be `ndim` instead of `ndims`, right?

unbohn commented:
Ok, the infrastructure runs without errors for me now. Will inspect results qualitatively now...

unbohn commented:
Qualitative results are interesting. Here are reflectance RGBs from running a classic single surface inversion (upper image) vs. the multisurface approach with two surface categories (lower imagel):

<img width="1242" height="1280" alt="emit20250327T212148_rfl" src="https://github.com/user-attachments/assets/b8be5c41-5591-4db4-a098-7cf921ff919e" />
<img width="1242" height="1280" alt="emit20250327T212148_rfl_multisurface" src="https://github.com/user-attachments/assets/aa543593-6e79-4d7d-a7b7-7082a4aee1ae" />

evan-greenbrg commented:
Action items:

- [ ] Resolve question about splitting .mat inputs for multisurface runs
- [ ] Test processing loads with new resource tracker

evan-greenbrg committed:
Adding raise error for multisurface run with input .mat surface file

evan-greenbrg committed:
Fixing config checks

evan-greenbrg commented:
Resource logs for multisurface run:

<img width="1066" height="631" alt="Screenshot 2025-09-09 at 8 49 39 AM" src="https://github.com/user-attachments/assets/5fe0cba6-5c9d-485e-8f4f-f9f4ca057f07" />
